lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Map.hs

```diff
@@ -0,0 +1,244 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE StrictData #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Cardano.Wallet.Deposit.Map
+    ( Map (..)
+    , K
+    , W
+    , type (^^^)
+    , Lookup
+    , Key (..)
+    , singletonMap
+    , openMap
+    , singletonPatched
+    , unPatch
+    , withMap
+    , forMap
+    , withPatched
+    , lookup
+    , lookupPatched
+    , forPatched
+    , openPatched
+    , At (..)
+    , forgetPatch
+    ) where
+
+import Prelude hiding
+    ( lookup
+    )
+
+import Data.Kind
+    ( Type
+    )
+import Data.Map.Monoidal.Strict
+    ( MonoidalMap (..)
+    )
+
+import qualified Data.Map.Monoidal.Strict as MonoidalMap
+
+-- | Infix form of MonoidalMap type
+type (^^^) = MonoidalMap
+
+infixr 5 ^^^
+
+-- | A phantom type for tuples of mappings from 'k' tpled with a spurious monoid
+-- 'w'. This is used to keep track of the patches applied to the map.
+data W (w :: Type) (k :: Type)
+
+-- | A phantom type for mappings from 'k'
+data K (k :: Type)
+
+-- | A nested monoidal map. Every nesting can also be patched with a monoid 'w'.
+data Map :: [Type] -> Type -> Type where
```

Comment:
It covers (obv) only acyclic graphs of joins, so a forest will be needed in case of full joins. Not sure I got your point. But if we'll reuse it we will find the time to evolve it.